src/Models/_OfflineModelBase.js

```diff
@@ -0,0 +1,215 @@
+        odef.resolve(updateResult);
+      }, function updateRelatedFailed(err) {
+        odef.reject(err);
+      });
+    }.bind(this), function updateFailed() {
+      // Fetching the doc/entity failed, so we will insert a new doc instead.
+      this.insertEntry(entry, options).then(function insertSuccess(insertResult) {
+        const odef = def;
+        this.saveRelatedEntries(entry, options).then( function insertRelatedSuccess() {
+          odef.resolve(insertResult);
+        }, function insertRelatedFailed(err) {
+          odef.reject(err);
+        });
+      }.bind(this), function insertFailed(err) {
+        def.reject(err);
+      });
+    }.bind(this));
+    return def.promise;
+  },
+  insertEntry: function insertEntry(entry, options) {
+    const store = this.getStore();
+    const def = new Deferred();
+    const doc = this.wrap(entry, options);
+    store.add(doc).then(function insertSuccess(result) {
+      def.resolve(result);
+    },
+    function insertFailed(err) {
+      def.reject('error inserting entity: ' + err);
+    });
+    return def.promise;
+  },
+  updateEntry: function updateEntity(entry, options) {
+    const store = this.getStore();
+    const def = new Deferred();
+    const entityId = this.getEntityId(entry, options);
+    this.getEntryDoc(entityId).then(function querySuccess(doc) {
+      const odef = def;
+      doc.entity = entry;
+      doc.modifyDate = moment().toDate();
+      doc.description = this.getEntityDescription(entry);
+      store.put(doc).then(function updateSuccess(result) {
+        odef.resolve(result);
+      }, function updateFailed(err) {
+        odef.reject('error updating entity: ' + err);
+      });
+    }.bind(this), function queryError(err) {
+      def.reject('entity not found to update:' + err);
+    });
+    return def.promise;
+  },
+  createEntry: function createEntry() {
+    const entry = {}; // need to dynamicly create Properties;
+    entry.Id = null;
+    entry.CreateDate = moment().toDate();
+    entry.ModifyDate = moment().toDate();
+    return entry;
+  },
+  deleteEntry: function deleteEntry(entityId) {
+    const def = new Deferred();
+    const store = this.getStore();
+    store.get(entityId).then((doc) => {
+      const odef = def;
+      store.remove(doc._id, doc._rev).then((result) => {
+        odef.resolve(result);
+      }, (err) => {
+        odef.reject(err);
+      });
+    }, (err) => {
+      def.reject(err);
+    });
+    return def.promise;
+  },
+  saveRelatedEntries: function saveRelatedEntries(parentEntry, options) {
+    const entries = (parentEntry && parentEntry.$relatedEntities) ? parentEntry.$relatedEntities : [];
+    const relatedPromises = [];
+    const def = new Deferred();
+    entries.forEach((related) => {
+      const model = App.ModelManager.getModel(related.entityName, MODEL_TYPES.OFFLINE);
+      if (model && related.entities) {
+        related.entities.forEach((relatedEntry) => {
```

Comment:
Instead of a forEach you can just use map, since saveEntry returns a promise:
`relatedPromises = related.entities.map(relatedEntry) => model.saveEntry(relatedEntry, options))`